File: exercise/scala/day13/src/main/scala/santamarket/model/ChristmasElf.scala
package santamarket.model

class ChristmasElf(val catalog: SantamarketCatalog):
  private val offers = scala.collection.mutable.Map[Product, Offer]()

  def addSpecialOffer(offerType: SpecialOfferType, product: Product, argument: Double): Unit =
    offers(product) = Offer(offerType, product, argument)

  def checksOutArticlesFrom(sleigh: ShoppingSleigh): Receipt =
    val receipt = Receipt()
    sleigh.getItems.foreach { pq =>
      val p = pq.product
      val quantity = pq.quantity
      val unitPrice = catalog.getUnitPrice(p)
      val price = quantity * unitPrice
      receipt.addProduct(p, quantity, unitPrice, price)
    }
    sleigh.handleOffers(receipt, offers.toMap, catalog)
    receipt

File: exercise/kotlin/day13/src/main/kotlin/santamarket/model/ChristmasElf.kt
package santamarket.model

class ChristmasElf(private val catalog: SantamarketCatalog) {
    private val offers = mutableMapOf<Product, Offer>()

    fun addSpecialOffer(offerType: SpecialOfferType, product: Product, argument: Double) {
        offers[product] = Offer(offerType, product, argument)
    }

    fun checksOutArticlesFrom(sleigh: ShoppingSleigh): Receipt {
        val receipt = Receipt()
        val productQuantities = sleigh.getItems()

        productQuantities.forEach { pq ->
            val product = pq.product
            val quantity = pq.quantity
            val unitPrice = catalog.getUnitPrice(product)
            val price = quantity * unitPrice
            receipt.addProduct(product, quantity, unitPrice, price)
        }

        sleigh.handleOffers(receipt, offers, catalog)
        return receipt
    }
}